### ISSUE TITLE
Error in writing Structure to cif file, when site species have "spin" property

### ISSUE THREAD

**Issue by kamronald** [materialsproject/pymatgen#3065]
Structure arose when enumerating magnetic structures, related to #3006 . When saving the enumerated structures, cif files are very different from the actual Structure objects generated by MagneticStructureEnumerator.
Writing to cif is fixed when the structure is modified such that spins are not defined, but replaced by defining a "magmom" site property.

**Comment by mkhorton** [comment]
Be aware that disordered magnetic materials are possible. I've never been entirely satisfied with the distinction between Species.spin and the magmom site property, but the former does allow a disordered magnetic material to be defined, while the latter does not.

**Comment by shyuep** [comment]
I would distinguish spin as an theoretical construct attached to an element. Eg a low spin Co3+ in octahedral has a spin of 0. Magmom is an actual calculated property. Eg an actual Co3+ in LiCoO2 can have a non zero magmom.

**Comment by mkhorton** [comment]
I remember this comment from you last time we discussed @shyuep :)

I think the reason I find it unsatisfying is simply that the site properties are not typed, and the use of "magmom" is mostly by our own convention.

**Comment by janosh** [comment]
Is this technically a bug? Not sure if `pymatgen` guarantees CIF round-trip identity?

@mkhorton Are you mostly concerned about not following a standard or site props being untyped? We could start typing by defining a `SiteProps` `TypedDict`:

```py
from typing import TypedDict

class SiteProps(TypedDict, total=true):
    magmom: list[float]
    charge: list[float]
```

`total=True` allows any key to be present in the dictionary, even if not specified in the type hint.

@kamronald Could you post a minimal repro [similar to this](https://github.com/materialsproject/pymatgen/issues/3006#issuecomment-1563093693)?

**Comment by kamronald** [comment]
Yes here is the repro. I have attached the CONTCAR and output structures below.

```
from pymatgen.core import Structure, PeriodicSite
from pymatgen.analysis.magnetism import MagneticStructureEnumerator

struc = Structure.from_file('CONTCAR')
mag_enum = MagneticStructureEnumerator(structure=struc, automatic=False, default_magmoms={'Li': 0.0, 'Mn': 4.0, 'O': 0.0},
                                       strategies=['ferromagnetic', 'antiferromagnetic'])
for i, struc in enumerate(mag_enum.ordered_structures):
    struc.to(f'enum_{i}.cif')

def replace_spins_w_magmom(structure):
    sites = []
    for spec, site in zip(structure.species, structure.sites):
        if 'spin' in spec._properties:
            spin = spec._properties['spin']
        else:
            spin = 0.0
        new_site = PeriodicSite(species=spec.symbol, coords=site.frac_coords, lattice=site.lattice,
                               properties={'magmom': spin})
        sites.append(new_site)
    return Structure.from_sites(sites)

for i, struc in enumerate(mag_enum.ordered_structures):
    magmom_struc = replace_spins_w_magmom(struc)
    magmom_struc.to(f'enum_mm_{i}.cif')
```
[mag_debug_strucs.zip](https://github.com/materialsproject/pymatgen/files/11761255/mag_debug_strucs.zip)

**Comment by kamronald** [comment]
I have found the error: adding the site "spin" property causes certain columns of data in the cif file to be "shifted". Specifically, the site x coordinate column contains the site occupancies, y coordinate column actually contains the x coordinates, z coordinate column contains the y coordinate, and occupancy coordinate contains the site occupancies. 
I have attached a screenshot of this data from enum_1.cif (which has the wrong columns) and the original enumerated structure. The info about the species spin actually takes up a column in the incorrect cif file, which likely leads to the error.

![enum_1_mm_cif_data](https://github.com/materialsproject/pymatgen/assets/35947132/57177c64-7e77-47d8-9b29-49afa70eef6a)
![enum_1_cif_data](https://github.com/materialsproject/pymatgen/assets/35947132/cbce9fb2-8af2-4bb6-b28e-d9c44061df7e)

**Comment by kamronald** [comment]
After changing a couple lines in CifWriter, the atomic locations in cif files are now correct, done in this commit: [([019dd8a](https://github.com/materialsproject/pymatgen/commit/019dd8a82e5bbbce6fbb15aaddc5fae772504dba))]
But then the spin info would now be lost. Could we add a separate column to include this data?
@shyuep  @janosh @mkhorton @mattmcdermott

**Comment by kamronald** [comment]
Actually, if the user wants to save magmom or spin info, they can do so by saving as MCIF. I just checked that this functionality works

**Comment by mkhorton** [comment]
More or less agree, but noting that during the mCIF specification discussions it was an open question whether they would add a scalar magmom attribute; last I checked, mCIF is correctly more intended for vector moments (eg from a non-collinear calculation), so the current mCIF writer cheats a little by choosing an arbitrary axis.For regular CIF files, we could equally just write out spin or other site properties under our own prefix. We actually reserved the “_mp” prefix for our own usage: https://www.iucr.org/cgi-bin/cifreserve.plOn Jun 15, 2023, at 15:31, Ronald Kam ***@***.***> wrote:﻿
Actually, if the user wants to save magmom or spin info, they can do so by saving as MCIF. I just checked that this functionality works
